HW2/P5/physics.pyx

```diff
@@ -91,41 +168,72 @@ cpdef update(FLOAT[:, ::1] XY,
         FLOAT *XY1, *XY2, *V1, *V2
         # SUBPROBLEM 4: uncomment this code.
-        # omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
+        omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
 
     assert XY.shape[0] == V.shape[0]
     assert XY.shape[1] == V.shape[1] == 2
 
     with nogil:
-        # bounce off of walls
+        ############## bounce off of walls ################################
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
         # scheduling.
+
+        ### 4 Threads
+        #for i in prange(count, schedule='static', chunksize=chunksize, num_threads=4):
+
+        ### Single Thread
         for i in range(count):
+
             for dim in range(2):
                 if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
                     ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
                     V[i, dim] *= -1
 
-        # bounce off of each other
+        ############### bounce off of each other ##########################
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
-        # scheduling.
+        # scheduling. 
+
+        ### 4 Threads
+        #for i in prange(count, schedule='static', chunksize=chunksize, num_threads=4):
+
+        ### Single Thread
         for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
 
-        # update positions
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, locks)
+
+        ############### update positions ##################################
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
         #    scheduling).
         # SUBPROBLEM 2: update the grid values.
+        # I do this within the same prange() (or range()) as the update step. 
+
+        ### 4 Threads
+        #for i in prange(count, schedule='static', chunksize=chunksize, num_threads=4): 
+
+        ### Single Thread       
         for i in range(count):
+
+            # If the ball is still within the grid, clear its position out of the grid before updating
+            if (XY[i,0]>=0 and XY[i,0]<=1) and (XY[i,1]>=0 and XY[i,1]<=1): 
+                x_grid = <unsigned int>(XY[i,0] / grid_spacing)
+                y_grid = <unsigned int>(XY[i,1] / grid_spacing)
+                Grid[x_grid, y_grid] = -1
+            # Update position
             for dim in range(2):
                 XY[i, dim] += V[i, dim] * t
-
-
+            # If the ball is still within the grid, update its Grid position after the update.  
+            if (XY[i,0]>=0 and XY[i,0]<=1) and (XY[i,1]>=0 and XY[i,1]<=1): 
+                x_grid = <unsigned int>(XY[i,0] / grid_spacing)
+                y_grid = <unsigned int>(XY[i,1] / grid_spacing)
+                Grid[x_grid, y_grid] = i
+ 
 def preallocate_locks(num_locks):
     cdef omp_lock_t *locks = get_N_locks(num_locks)
     assert 0 != <uintptr_t> <void *> locks, "could not allocate locks"
```

Comment:
I did similar things. I think most of the enhancement is due to optimizing the algorithm instead of parallel tools for this problem.